### PR TITLE
First S3 store implementation

### PR DESCRIPTION
**Pull request by filosganga**

### CODE REVIEW

File: src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala [hunk 1]
```diff
@@ -1,9 +1,84 @@
...
+  case class Config(bucketName: Bucket,
+                    region: Region = Region.`eu-west-1`)
+
+  trait Keys[F[_]] {
+    def get(commId: CommId, traceToken: TraceToken): F[Key]
+  }
+
+  class RandomSuffixKeys[F[_]: Applicative] extends Keys[F] {
+    override def get(commId: CommId, traceToken: TraceToken): F[Key] = {
+      Key(s"${commId}-${UUID.randomUUID()}").pure[F]
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala]: `randomUUID` is side-effectful. I've let it slide in tests, but not here :P

```scala
object Keys {
  def random[F[_]: Sync] = new Keys[F] {
    def get(commId: CommId, traceToken: TraceToken): F[Key] = 
       F.delay(UUID.randomUUID()).map { uuid =>
           Key(s"${commId}-${UUID.randomUUID()}")
       }
}
```

File: src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala [hunk 2]
```diff
@@ -1,9 +1,84 @@
...
+      } else {
+        s"s3-${region.value}.amazonaws.com"
+      }
+
+      for {
+        key <- keys.get(commId, traceToken)
+        result <- s3.putObject(config.bucketName, key, content, Map("comm-id"->commId, "trace-token"->traceToken))
+          .map { resultOrError =>
+            resultOrError
+              .leftMap(identity[Throwable])
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala]: this whole thing should be `.leftWiden`, if the inference gods play nicely

File: src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala [hunk 3]
```diff
@@ -1,9 +1,84 @@
...
+              // TODO Should be able to use *> ?
+              .flatMap { _ =>
+              Uri.fromString(s"https://${bucketName.name}.${s3Domain}/${key.value}")
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala]: yep, `>>` or `*>` should work here

File: src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala [hunk 4]
```diff
@@ -1,9 +1,86 @@
...
+  def stream[F[_]: Effect](config: Config): Stream[F, Store[F]] = {
+    Http1Client.stream[F]().map { client =>
+      apply(
+        new S3[F](client, CredentialsProvider.default[F], config.region),
+        config,
+        new RandomSuffixKeys[F])
+    }
+  }
+
+  def apply[F[_]: Sync](s3: S3[F], config: Config, keys: Keys[F]): Store[F] = new Store[F] {
```
Comment (laurence-bird) [inline src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala]: is there a way we can unit test this? would be good to have this covered

File: src/main/scala/com/ovoenergy/comms/composer/v2/Hash.scala [hunk 5]
```diff
@@ -25,11 +28,9 @@ trait Hashable[A] {
 
 object Hashable {
 
-  // TODO: remove side effectful operation
-  val messageDigest = MessageDigest.getInstance("MD5")
-
   implicit val hashableSms: Hashable[OrchestratedSMSV3] = new Hashable[OrchestratedSMSV3]() {
     def hash(a: OrchestratedSMSV3): String = {
+      val messageDigest = MessageDigest.getInstance("MD5")
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/composer/v2/Hash.scala]: ehm, the side effect is still there, so the `TODO` is still valid

File: src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala (hunk 4 above)
Comment (filosganga) [reply src/main/scala/com/ovoenergy/comms/composer/v2/Store.scala]: Sorted offline

**Pull request merged**